### PR TITLE
Add config option to forbid deleting user account

### PR DESCRIPTION
**Pull request by aviav**
In some cases it might make sense to prevent the last user of an account from loosing his ability to sign in. This configuration option leaves the details of this policy to the host application.

### CODE REVIEW

File: admins/pageflow/user.rb [hunk 1]
```diff
@@ -120,7 +120,8 @@ module Pageflow
 
     collection_action 'delete_me', :title => I18n.t('pageflow.admin.users.account'), :method => [:get, :delete] do
       if request.delete?
-        if current_user.destroy_with_password(params.require(:user)[:current_password])
+        if Pageflow.config.authorize_user_deletion.call == true &&
```
Comment (tf) [inline admins/pageflow/user.rb]: Let's use `authorized?` here and move the call of the configured lambda to the `AbilityMixin` (and later on to a policy).

File: app/views/admin/users/me.html.erb [hunk 2]
```diff
@@ -15,7 +15,11 @@
...
   <%= form.actions %>
   <p>
-    <%= t('pageflow.admin.users.me.delete_account_hint') %>
-    <%= link_to(t('pageflow.admin.users.me.delete_account'), delete_me_admin_users_path, :class => 'delete_account') %>
+    <% if Pageflow.config.authorize_user_deletion.call == true %>
+      <%= t('pageflow.admin.users.me.delete_account_hint') %>
+        <%= link_to(t('pageflow.admin.users.me.delete_account'), delete_me_admin_users_path, :class => 'delete_account') %>
+    <% else %>
+        <%= Pageflow.config.authorize_user_deletion.call %>
+    <% end %>
```
Comment (tf) [inline app/views/admin/users/me.html.erb]: Let's move this to a helper rendering a partial. That way we can also avoid calling the lambda multiple times as it can potentially perform expensive queries. Also let's apply different class names to the surrounding paragraph tag in each case. That way we can style the message to include a warning icon etc when deleting is forbidden.

File: lib/pageflow/configuration.rb [hunk 3]
```diff
@@ -228,6 +228,8 @@ class Configuration
     attr_accessor :default_author_meta_tag
     attr_accessor :default_publisher_meta_tag
 
+    attr_accessor :authorize_user_deletion
+
```
Comment (tf) [inline lib/pageflow/configuration.rb]: Please add inline help describing the option and the lambda signature. Also add a `@since edge` doc tag. That will automatically be rewritten to the correct version when the gem is released.

File: admins/pageflow/user.rb (hunk 1 above)
Comment (tf) [inline admins/pageflow/user.rb]: Also let's always pass a `user` parameter. Otherwise there is no way for implementors to know which user we are talking about.

File: spec/controllers/admin/users_controller_spec.rb [hunk 4]
```diff
@@ -146,5 +146,27 @@ module Pageflow
         expect(user.reload.account).to eq(other_account)
       end
     end
+
+    describe '#delete_me' do
+      it 'allows to destroy the user when authorize_user_deletion true' do
```
Comment (tf) [inline spec/controllers/admin/users_controller_spec.rb]: I'd rephrase that to "by default". That way we do not leak to much implementation detail into spec description. The actual default setup might change in the future.

File: spec/controllers/admin/users_controller_spec.rb [hunk 5]
```diff
@@ -146,5 +146,27 @@ module Pageflow
...
+          delete(:delete_me, user: {current_password: '@qwert123'})
+        end.to change { User.count }
+      end
+
+      it 'does not allow to destroy the user when authorize_user_deletion non-true' do
+        user = create(:user, password: '@qwert123')
+        account_users_count = user.account.users.length
+        sign_in(user)
+        Pageflow.config.authorize_user_deletion =
+          lambda { account_users_count > 1 ? true : 'Last user on account. Permission denied' }
```
Comment (tf) [inline spec/controllers/admin/users_controller_spec.rb]: Normally I'd say, simply return `false` to make the test as simple as possible. On the other hand, we need to test that the block gets passed a user record. So performing an actual check inside the lambda is one way of expressing that while also demonstrating real world usage. The alternative would be to pass a mock and explicitly expect that `call` is invoked with given arguments.

File: admins/pageflow/user.rb (hunk 1 above)
Comment (aviav) [inline admins/pageflow/user.rb]: Fixed

File: app/views/admin/users/me.html.erb (hunk 2 above)
Comment (aviav) [inline app/views/admin/users/me.html.erb]: Done

File: lib/pageflow/configuration.rb (hunk 3 above)
Comment (aviav) [inline lib/pageflow/configuration.rb]: Added.

File: spec/controllers/admin/users_controller_spec.rb (hunk 4 above)
Comment (aviav) [inline spec/controllers/admin/users_controller_spec.rb]: Rephrased.

File: app/helpers/pageflow/admin/users_helper.rb [hunk 6]
```diff
@@ -4,6 +4,14 @@ module UsersHelper
       def collection_for_user_roles
         User.roles_accessible_by(current_ability).index_by { |role| t(role, :scope => 'pageflow.admin.users.roles') }
       end
+
+      def delete_me_para
+        if Pageflow.config.authorize_user_deletion.call(current_user) == true
+          render('may_delete')
+        else
+          render('cannot_delete')
+        end
```
Comment (tf) [inline app/helpers/pageflow/admin/users_helper.rb]: * Let's save the result of `call` here and pass it to the partial as local variable. That way we do not call the lambda multiple times and do not risk calling it without parameters in some cases (see below)
* To catch such errors, let's add a test which covers the two cases and makes sure the returned string is rendered.
* Since all helpers share a global scope, helper method names should include the name of their module, i.e. `delete_own_user_section` etc.

File: lib/pageflow/ability_mixin.rb [hunk 7]
```diff
@@ -35,6 +35,10 @@ def pageflow_default_abilities(user)
 
       can :view, [Admin::MembersTab, Admin::RevisionsTab]
 
+      can :delete_own, ::User do |current_user|
+        Pageflow.config.authorize_user_deletion.call(current_user) == true
+      end
+
```
Comment (tf) [inline lib/pageflow/ability_mixin.rb]: No need to pass in `current_user` here. Let's call the ability `:delete_own_user`.

File: spec/controllers/admin/users_controller_spec.rb [hunk 8]
```diff
@@ -146,5 +146,32 @@ module Pageflow
...
+            else
+              'Last user on account. Permission denied'
+            end
+          end
+
+        expect do
+          delete(:delete_me, user: {current_password: '@qwert123'})
+        end.not_to change { User.count }
+      end
+    end
```
Comment (tf) [inline spec/controllers/admin/users_controller_spec.rb]: I like the real life example here.

File: app/helpers/pageflow/admin/users_helper.rb (hunk 6 above)
Comment (aviav) [inline app/helpers/pageflow/admin/users_helper.rb]: Fixed

File: lib/pageflow/ability_mixin.rb (hunk 7 above)
Comment (aviav) [inline lib/pageflow/ability_mixin.rb]: Yes need. With calls of the form `authorized?(:delete_own_user, ::User)` CanCan will ignore the block. See `can?(:read, Account)`

File: app/views/admin/users/me.html.erb [hunk 9]
```diff
@@ -14,8 +14,5 @@
     <%= form.input :password_confirmation %>
   <% end %>
   <%= form.actions %>
-  <p>
-    <%= t('pageflow.admin.users.me.delete_account_hint') %>
-    <%= link_to(t('pageflow.admin.users.me.delete_account'), delete_me_admin_users_path, :class => 'delete_account') %>
-  </p>
+    <%= delete_own_user_section %>
```
Comment (tf) [inline app/views/admin/users/me.html.erb]: Indentation looks off here.

File: spec/features/editor/deleting_account_spec.rb [hunk 10]
```diff
@@ -11,4 +11,23 @@
...
+          true
+        else
+          'Deletion is only possible for members of a dedicated account'
+        end
+      end
+
+    visit('/admin/users/me')
+
+    expect(page).to have_content(Dom::Admin::ProfileForm.first.cannot_delete_comment.text)
+  end
```
Comment (tf) [inline spec/features/editor/deleting_account_spec.rb]: I think this test would be more helpful (and faster) as a helper spec. We are also mostly covering happy paths in a lot of the other feature specs.

**Comment by tf** [comment]
Starting to look really good. Please squash the commits into a single one when you are done.

File: app/views/admin/users/me.html.erb (hunk 9 above)
Comment (aviav) [inline app/views/admin/users/me.html.erb]: Fixed

File: spec/features/editor/deleting_account_spec.rb (hunk 10 above)
Comment (aviav) [inline spec/features/editor/deleting_account_spec.rb]: Agreed, moved non-happy path to helper spec

**Comment by aviav** [comment]
Squashed all commits into one

File: spec/helpers/pageflow/admin/users_helper_spec.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+          expect(helper).to receive(:delete_me_admin_users_path).and_return('#')
+
+          section = helper.delete_own_user_section
+
+          expect(section).to have_selector('.may_delete')
+          expect(section).not_to have_selector('.cannot_delete')
+        end
+
+        it 'renders cannot_delete if authorize_user_deletion gives any string' do
+          Pageflow.config.authorize_user_deletion = lambda { |_user| 'true' }
```
Comment (tf) [inline spec/helpers/pageflow/admin/users_helper_spec.rb]: This is confusing. The quotes are hard to spot. Return a string like "Not allowed" and also query for it inside the resulting HTML.

Comment (tf) [inline spec/helpers/pageflow/admin/users_helper_spec.rb]: Maybe again use the user in the lambda to ensure it is passed.

File: spec/helpers/pageflow/admin/users_helper_spec.rb [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+module Pageflow
+  module Admin
+    describe UsersHelper do
+      describe '#delete_own_user_section' do
+        it 'renders may_delete if authorize_user_deletion gives true' do
+          Pageflow.config.authorize_user_deletion = lambda { |_user| true }
+
+          expect(helper).to receive(:current_user).and_return(create(:user))
+          expect(helper).to receive(:delete_me_admin_users_path).and_return('#')
```
Comment (tf) [inline spec/helpers/pageflow/admin/users_helper_spec.rb]: Let's use `allow` here. Try prefixing the route helpers in the template with `main_app`. Then stubbing is not required. Compare app/views/components/pageflow/admin/grouped_folder_list.rb

File: app/views/pageflow/admin/users/_may_delete.html.erb [hunk 13]
```diff
@@ -0,0 +1,6 @@
+<p class="may_delete">
+  <%= t('pageflow.admin.users.me.delete_account_hint') %>
+    <%= link_to(t('pageflow.admin.users.me.delete_account'),
+                delete_me_admin_users_path,
+                :class => 'delete_account') %>
```
Comment (tf) [inline app/views/pageflow/admin/users/_may_delete.html.erb]: Where is Hound? ;)

Comment (aviav) [inline app/views/pageflow/admin/users/_may_delete.html.erb]: Is this stylistically incorrect? I do not see anything wrong here

Comment (tf) [inline app/views/pageflow/admin/users/_may_delete.html.erb]: Hash syntax?

Comment (aviav) [inline app/views/pageflow/admin/users/_may_delete.html.erb]: Hm, where is Hound indeed? https://twitter.com/houndci does not report any malfunction

Comment (tf) [inline app/views/pageflow/admin/users/_may_delete.html.erb]: I guess they do not lint templates?

Comment (aviav) [inline app/views/pageflow/admin/users/_may_delete.html.erb]: In all fairness, they do not explicitly advertise checking embedded Ruby on their home page https://houndci.com/

File: spec/helpers/pageflow/admin/users_helper_spec.rb (hunk 11 above)
Comment (aviav) [inline spec/helpers/pageflow/admin/users_helper_spec.rb]: Done

File: spec/helpers/pageflow/admin/users_helper_spec.rb (hunk 12 above)
Comment (aviav) [inline spec/helpers/pageflow/admin/users_helper_spec.rb]: Done
